pkg/operation/etcdencryption/encryptionconfiguration.go

```diff
@@ -0,0 +1,178 @@
+//       keys:
+//       - name: key1559747207815249000
+//         secret: Y8LEzbtK/2mdXrw6W/faAxNLu+mTCmcQeWojShAJGEg=
+//   - identity: {}
+//   resources:
+//     metadata:
+//   - secrets
+func NewPassiveConfiguration() (*apiserverconfigv1.EncryptionConfiguration, error) {
+	key, err := NewEncryptionKey()
+	if err != nil {
+		return nil, err
+	}
+	return &apiserverconfigv1.EncryptionConfiguration{
+		TypeMeta: v1.TypeMeta{
+			Kind:       ecKind,
+			APIVersion: ecAPIVersion,
+		},
+		Resources: []apiserverconfigv1.ResourceConfiguration{
+			{
+				Resources: ecEncryptedResources,
+				Providers: []apiserverconfigv1.ProviderConfiguration{
+					{Identity: &apiserverconfigv1.IdentityConfiguration{}},
+					{AESCBC: &apiserverconfigv1.AESConfiguration{
+						Keys: []apiserverconfigv1.Key{
+							*key,
+						},
+					}},
+				},
+			},
+		},
+	}, nil
+}
+
+// CreateFromYAML creates a new configuration from a YAML String as Byte Array.
+func CreateFromYAML(ecYamlBytes []byte) (*apiserverconfigv1.EncryptionConfiguration, error) {
+	scheme := runtime.NewScheme()
+	codecs := serializer.NewCodecFactory(scheme)
+	utilruntime.Must(apiserverconfigv1.AddToScheme(scheme))
+	ec := &apiserverconfigv1.EncryptionConfiguration{}
+	_, _, err := codecs.UniversalDecoder().Decode(ecYamlBytes, nil, ec)
+	if err != nil {
+		return nil, fmt.Errorf("error while decoding EncryptionConfiguration from yamlArray: %v", err)
+	}
+	return ec, nil
+}
+
+// ToYAML Creates a YAML representation of the EncryptionConfiguration.
+func ToYAML(ec *apiserverconfigv1.EncryptionConfiguration) ([]byte, error) {
+	scheme := runtime.NewScheme()
+	codecs := serializer.NewCodecFactory(scheme)
+	err := apiserverconfigv1.AddToScheme(scheme)
+	if err != nil {
+		return nil, fmt.Errorf("error while preparing parsing of EncryptionConfiguration: %v", err)
+	}
+	serializer := json.NewYAMLSerializer(json.DefaultMetaFactory, scheme, scheme)
+	encoder := codecs.EncoderForVersion(serializer, apiserverconfigv1.SchemeGroupVersion)
+	ecYamlBytes, err := runtime.Encode(encoder, ec)
+	if err != nil {
+		return nil, fmt.Errorf("error while parsing EncryptionConfiguration: %v", err)
+	}
+	return ecYamlBytes, nil
+}
+
+// Validate checks whether the configuration is consistent.
+// Consistency checks include the following:
+// Has the secret etcd-encryption-secret a data item named encryption-configuration.yaml?
+// Consists the list of encryption providers of exactly the following 2 entries: "identity", "aescbc")?
+// Is a key present with a sensible timestamp in its name and a sensible value?
+// Is the additional data item encryption-metadata.yaml available and valid?
+// Is the secret made available as a volume and via a volume mount to the API server pod in the shoot's API server deployment?
+// Is the file encryption-configuration.yaml referenced in a startup parameter --encryption-provider-config of the shoot's API server?
+func Validate(ec *apiserverconfigv1.EncryptionConfiguration) error {
```

Comment:
Neither customers nor operators are supposed to touch the EncryptionConfiguration ... this is different for the shoots that you mention. I fail to see why this change is required.

Reply:
Also, the ErrorList seems to me to make sense only if the errors are somehow independent whereas for the EncryptionConfiguration checks, one builds upon the next ... e.g. if kind or version do not match, there is no point to continue checking further since the resource is no EncryptionConfiguration. 

What I could do is to add a template for a correct EncyrptionConfiguration to the error message to hind an operator (who absolutely wants to play with fire) how the EncryptionConfiguration ought to look like here. 

Do you want this?

Reply:
No, nobody should manual modify this anyhow.

Reply:
I still vote for having the validation as in any other k8s resource - for me, having a proper message that shows me directly what's the issue is very important. Also, from your code, it shouldn't be hard to go there.
> Customers cannot modify the EncryptionConfiguration, operators are not supposed to touch the EncryptionConfiguration ... in my opinion this is quite different for the shoots that you mention. Thus I fail to see why this change is required.

Yes, they cannot modify it, but if there is an error at any point in time, having this done correctly will save a lot of time for the guy who has to do the ops.
> Also, the ErrorList seems to me to make sense only if the errors are somehow independent whereas for the EncryptionConfiguration checks, one builds upon the next ... e.g. if kind or version do not match, there is no point to continue checking further since the resource is no EncryptionConfiguration.

Of course you could exit immediately once you see the first error, but having all errors at one glance helps to reduce feedback time drastically by seeing all mistakes you've possibly done.

Last but not least, the whole code will become way less entangled than it is currently. Right now, the cyclomatic complexity of this method is too high, separating it correctly will also help maintaining this in the future.

Reply:
> I still vote for having the validation as in any other k8s resource - for me, having a proper message that shows me directly what's the issue is very important. Also, from your code, it shouldn't be hard to go there.

Also for me, having a proper message is important. You could see this in the error texts of the `Validate` function. 

> > Customers cannot modify the EncryptionConfiguration, operators are not supposed to touch the EncryptionConfiguration ... in my opinion this is quite different for the shoots that you mention. Thus I fail to see why this change is required.
> 
> Yes, they cannot modify it, but if there is an error at any point in time, having this done correctly will save a lot of time for the guy who has to do the ops.

Where should the faulty configuration come from, if the customer cannot modify it? Starting to repeat myself here, but I could add a template for a correct EncryptionConfiguration so that a zealous operator sees how it should look like if she or he messed it up for some unfathomable reason. Rafael didn't think it necessary. 

> > Also, the ErrorList seems to me to make sense only if the errors are somehow independent whereas for the EncryptionConfiguration checks, one builds upon the next ... e.g. if kind or version do not match, there is no point to continue checking further since the resource is no EncryptionConfiguration.
> 
> Of course you could exit immediately once you see the first error, but having all errors at one glance helps to reduce feedback time drastically by seeing all mistakes you've possibly done.

The reason why I am resisting here is, that the checks which I do in the `Validate` function are NOT independent from each other but build on each other instead. Thus, I cannot simply go as  "the first error happened, but let's just assume this was not the case" 

> Last but not least, the whole code will become way less entangled than it is currently. Right now, the cyclomatic complexity of this method is too high, separating it correctly will also help maintaining this in the future.

I do not see this function as being "entangled"... I go through the structure top-down and check all elements if they match the expectations.

Reply:
I see that they are dependent on each other, however, if you put them into blocks like `ValidateResources` using `field.ErrorList`, then
* it becomes much easier to see how the structure of the configuration is validated
* it is way easier to construct meaningful errors going to the path of the invalid field using the `field.<>` error constructors
* you can write proper tests ensuring that exactly the errors you expected are thrown.

Reply:
Ok, more nice-to-have fixes .. at least from my point of view .. anyway .. 

As desired I have now added a new validation package and moved all validation functions there. 

Also I changed the validations as desired to work with the `field.ErrorList` and splitted everything into many more validation functions. All for something that nobody should ever touch manually ... 

Question: I call my validation from `LoadFromYAML`. Do I assume correctly that I have to change now all APIs up to CreateOrUpdateEtcdEncryptionConfiguration which is called in the reconciliation? Does the reconsiliation cope with the `field.ErrorList` as well as the `error` ?